File: 0519/택배 배달과 수거하기.java
class Solution {
    
    static int cap; // 트럭 용량
    static int n; // 집의 개수
    static int[] deliveries; // 각 집에 배달할 택배 상자 개수
    static int[] pickups; // 각 집에서 수거할 택배 상자 개수
    static int dCursor;
    static int pCursor;
    
    public long solution(int cap, int n, int[] deliveries, int[] pickups) {
        init(cap, n, deliveries, pickups);
        dCursor = findNearValidLeft(deliveries, deliveries.length - 1);
        pCursor = findNearValidLeft(pickups, pickups.length - 1);
        
        long answer = 0;
        while (dCursor > -1 || pCursor > -1) {
            int dist = Math.max(dCursor, pCursor) + 1;
            answer += dist * 2;
            
            dCursor = process(deliveries, dCursor);
            pCursor = process(pickups, pCursor);
        }
        
        return answer;
    }
    
    void init(int cap, int n, int[] deliveries, int[] pickups) {
        this.cap = cap;
        this.n = n;
        this.deliveries = deliveries;
        this.pickups = pickups;
    }
    
    int findNearValidLeft(int[] arr, int start) {
        int cursor = start;
        while (cursor > -1 && arr[cursor] == 0) {
            cursor--;
        }
        
        return cursor;
    }
    
    int process(int[] arr, int cursor) {
        if (cursor == -1) {
            return -1;
        }
        
        for (int i=0; i<cap; i++) {
            arr[cursor]--;
            if (arr[cursor] == 0) {
                cursor = findNearValidLeft(arr, cursor);
                if (cursor == -1) {
                    return -1;
                }
            }
        }
        return cursor;
    }
}

File: 0519/택배 배달과 수거하기.py
def solution(cap, n, deliveries, pickups):
    i = n - 1
    j = n - 1
    result = 0

    while i > -1 or j > -1:
        deli_cap = 0
        pick_cap = 0

        deli_r = 0
        pick_r = 0

        while deli_cap < cap and i > -1:
            tmp = deliveries[i]

            if tmp:
                deli_r = max(deli_r, i)

                if deli_cap + tmp <= cap:
                    deli_cap += tmp
                    deliveries[i] = 0
                    i -= 1
                    
                else:
                    r = cap - deli_cap
                    deli_cap += r
                    deliveries[i] -= r
                    break
            else:
                i -= 1

        while pick_cap < cap and j > -1:
            tmp = pickups[j]
            
            if tmp:
                pick_r = max(pick_r, j)

                if pick_cap + tmp <= cap:
                    pick_cap += tmp
                    pickups[j] = 0
                    j -= 1
                else:
                    r = cap - pick_cap
                    pick_cap += r
                    pickups[j] -= r
                    break
            else:
                j -= 1
                
        if deli_r > -1 or pick_r > -1:
            if deli_r == 0 and pick_r == 0 and deli_cap == 0 and pick_cap == 0:
                result += 0
            else:
                result += (max(deli_r, pick_r) + 1) * 2
    
    return result
